roxie/udplib/udptrr.cpp

```diff
@@ -270,71 +710,118 @@ class CReceiveManager : implements IReceiveManager, public CInterface
             }
         }
 
-    };
+        // code to track the number of permits - all functions are called from the flow control thread, except for decPermit() from the data thread
+        // need to be careful about concurent modifications. The exact number isn't critical, but
+        // we should never return a -ve number.  Simplest to implement by checking in getNumReserved() rather than using a cas in decPermit()
+        // How many permits outstanding for a given flowSeq?
+        inline unsigned getNumReserved(unsigned flowSeq) const
+        {
+            for (unsigned i=0; i < MAX_PERMITS; i++)
+            {
+                if (permits[i].flowSeq == flowSeq)
+                    return permits[i].getNumReserved();
+            }
+            return 0;
+        }
 
-    class SenderList
-    {
-        UdpSenderEntry *head = nullptr;
-        UdpSenderEntry *tail = nullptr;
-        unsigned numEntries = 0;
+        //Total permits outstanding for the sender
```

Comment:
Total reservations, I think. There is scope for confusion when using the term 'permit' between an active permission to send and the number of slots that the permit grants...